File: task-9127bb/environment/vm.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Value struct {
	IsStr bool
	Num   int64
	Str   string
}

func (v Value) String() string {
	if v.IsStr {
		return fmt.Sprintf("%q", v.Str)
	}
	return strconv.FormatInt(v.Num, 10)
}

type Insn struct {
	Op  string
	Arg string
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: vm [--trace] <file.asm>\n")
		fmt.Fprintf(os.Stderr, "       vm --help\n")
		os.Exit(1)
	}

	trace := false
	file := ""

	for i := 1; i < len(os.Args); i++ {
		switch os.Args[i] {
		case "--help", "-h":
			printHelp()
			return
		case "--trace":
			trace = true
		default:
			if file != "" {
				fmt.Fprintf(os.Stderr, "Error: unexpected argument: %s\n", os.Args[i])
				os.Exit(1)
			}
			file = os.Args[i]
		}
	}

	if file == "" {
		fmt.Fprintf(os.Stderr, "Error: no input file specified\n")
		os.Exit(1)
	}

	insns, labels := load(file)
	run(insns, labels, trace)
}

func printHelp() {
	help := `vm - Stack-based virtual machine

Usage:
  vm <file.asm>              Execute an assembly program
  vm --trace <file.asm>      Execute with instruction-level tracing
  vm --help                  Show this help message

Tracing:
  When --trace is enabled, each instruction is logged to stderr before
  execution, showing the program counter, instruction, and current stack
  state. A summary of total instructions executed is printed at the end.
  Program output (stdout) is unaffected by tracing.

  Trace format per line:
    [pc] INSTRUCTION              stack: [values...]

  At end of execution:
    --- N instructions executed ---

Exit codes:
  0   Successful execution (HALT reached or end of program)
  1   Error (file not found, parse error, runtime error)`
	fmt.Println(help)
}

func load(path string) ([]Insn, map[string]int) {
	f, err := os.Open(path)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
	defer f.Close()

	var insns []Insn
	labels := map[string]int{}
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		parts := strings.SplitN(line, " ", 2)
		op := parts[0]
		arg := ""
		if len(parts) > 1 {
			arg = parts[1]
		}
		if op == "LABEL" {
			labels[arg] = len(insns)
		}
		insns = append(insns, Insn{Op: op, Arg: arg})
	}
	return insns, labels
}

func formatStack(stack []Value) string {
	if len(stack) == 0 {
		return "[]"
	}
	parts := make([]string, len(stack))
	for i, v := range stack {
		parts[i] = v.String()
	}
	return "[" + strings.Join(parts, ", ") + "]"
}

func run(insns []Insn, labels map[string]int, trace bool) {
	var stack []Value
	vars := map[string]Value{}
	reader := bufio.NewReader(os.Stdin)
	pc := 0
	traceCount := 0

	push := func(v Value) {
		stack = append(stack, v)
	}

	pop := func() Value {
		if len(stack) == 0 {
			fmt.Fprintf(os.Stderr, "Stack underflow at pc=%d\n", pc)
			os.Exit(1)
		}
		v := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		return v
	}

	for pc < len(insns) {
		insn := insns[pc]

		if trace {
			insnStr := insn.Op
			if insn.Arg != "" {
				insnStr += " " + insn.Arg
			}
			fmt.Fprintf(os.Stderr, "[%3d] %-25s stack: %s\n",
				pc, insnStr, formatStack(stack))
			traceCount++
		}

		switch insn.Op {
		case "LABEL":
			pc++
		case "PUSH_INT":
			n, _ := strconv.ParseInt(insn.Arg, 10, 64)
			push(Value{Num: n})
			pc++
		case "PUSH_STR":
			push(Value{IsStr: true, Str: insn.Arg})
			pc++
		case "LOAD":
			v, ok := vars[insn.Arg]
			if !ok {
				fmt.Fprintf(os.Stderr, "Undefined variable: %s at pc=%d\n", insn.Arg, pc)
				os.Exit(1)
			}
			push(v)
			pc++
		case "STORE":
			vars[insn.Arg] = pop()
			pc++
		case "DROP":
			pop()
			pc++
		case "DUP":
			v := pop()
			push(v)
			push(v)
			pc++
		case "SWAP":
			b, a := pop(), pop()
			push(b)
			push(a)
			pc++
		case "ADD":
			b, a := pop(), pop()
			push(Value{Num: a.Num + b.Num})
			pc++
		case "SUB":
			b, a := pop(), pop()
			push(Value{Num: a.Num - b.Num})
			pc++
		case "MUL":
			b, a := pop(), pop()
			push(Value{Num: a.Num * b.Num})
			pc++
		case "DIV":
			b, a := pop(), pop()
			if b.Num == 0 {
				fmt.Fprintf(os.Stderr, "Division by zero at pc=%d\n", pc)
				os.Exit(1)
			}
			push(Value{Num: a.Num / b.Num})
			pc++
		case "MOD":
			b, a := pop(), pop()
			if b.Num == 0 {
				fmt.Fprintf(os.Stderr, "Modulo by zero at pc=%d\n", pc)
				os.Exit(1)
			}
			push(Value{Num: a.Num % b.Num})
			pc++
		case "NEG":
			v := pop()
			push(Value{Num: -v.Num})
			pc++
		case "NOT":
			v := pop()
			if v.Num == 0 {
				push(Value{Num: 1})
			} else {
				push(Value{Num: 0})
			}
			pc++
		case "CMP_EQ":
			b, a := pop(), pop()
			if a.Num == b.Num {
				push(Value{Num: 1})
			} else {
				push(Value{Num: 0})
			}
			pc++
		case "CMP_NE":
			b, a := pop(), pop()
			if a.Num != b.Num {
				push(Value{Num: 1})
			} else {
				push(Value{Num: 0})
			}
			pc++
		case "CMP_LT":
			b, a := pop(), pop()
			if a.Num < b.Num {
				push(Value{Num: 1})
			} else {
				push(Value{Num: 0})
			}
			pc++
		case "CMP_LE":
			b, a := pop(), pop()
			if a.Num <= b.Num {
				push(Value{Num: 1})
			} else {
				push(Value{Num: 0})
			}
			pc++
		case "CMP_GT":
			b, a := pop(), pop()
			if a.Num > b.Num {
				push(Value{Num: 1})
			} else {
				push(Value{Num: 0})
			}
			pc++
		case "CMP_GE":
			b, a := pop(), pop()
			if a.Num >= b.Num {
				push(Value{Num: 1})
			} else {
				push(Value{Num: 0})
			}
			pc++
		case "PRINT":
			v := pop()
			if v.IsStr {
				fmt.Print(v.Str)
			} else {
				fmt.Print(v.Num)
			}
			pc++
		case "PRINTLN":
			v := pop()
			if v.IsStr {
				fmt.Println(v.Str)
			} else {
				fmt.Println(v.Num)
			}
			pc++
		case "READ_INT":
			line, err := reader.ReadString('\n')
			if err != nil && line == "" {
				fmt.Fprintf(os.Stderr, "Read error: %v\n", err)
				os.Exit(1)
			}
			line = strings.TrimSpace(line)
			n, err := strconv.ParseInt(line, 10, 64)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Invalid integer input: %s\n", line)
				os.Exit(1)
			}
			push(Value{Num: n})
			pc++
		case "JUMP":
			idx, ok := labels[insn.Arg]
			if !ok {
				fmt.Fprintf(os.Stderr, "Unknown label: %s at pc=%d\n", insn.Arg, pc)
				os.Exit(1)
			}
			pc = idx
		case "JUMP_TRUE":
			v := pop()
			if v.Num != 0 {
				pc = labels[insn.Arg]
			} else {
				pc++
			}
		case "JUMP_FALSE":
			v := pop()
			if v.Num == 0 {
				pc = labels[insn.Arg]
			} else {
				pc++
			}
		case "HALT":
			if trace {
				fmt.Fprintf(os.Stderr, "\n--- %d instructions executed ---\n", traceCount)
			}
			return
		default:
			fmt.Fprintf(os.Stderr, "Unknown instruction: %s at pc=%d\n", insn.Op, pc)
			os.Exit(1)
		}
	}

	if trace {
		fmt.Fprintf(os.Stderr, "\n--- %d instructions executed ---\n", traceCount)
	}
}

File: task-9127bb/environment/vm.py
#!/usr/bin/env python3
"""
Stack-based Virtual Machine for a GoAWK-inspired instruction set.

"""
import sys


class VMError(Exception):
    pass


def parse_program(text):
    """Parse assembly text into list of (opcode, arg) tuples."""
    instructions = []
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        parts = line.split(None, 1)
        op = parts[0]
        arg = None
        if len(parts) > 1:
            raw = parts[1]
            if raw.startswith('"') and raw.endswith('"'):
                arg = raw[1:-1]
                arg = arg.replace('\\n', '\n').replace('\\t', '\t')
                arg = arg.replace('\\\\"', '"').replace('\\\\', '\\')
            else:
                arg = raw
        instructions.append((op, arg))
    return instructions


class VM:
    def __init__(self, instructions, stdin_data=""):
        self.insns = instructions
        self.stack = []
        self.vars = {}
        self.pc = 0
        self.labels = {}
        self.output = []
        self.exec_count = 0
        self.halted = False

        clean = stdin_data.rstrip('\n')
        self.stdin_lines = clean.split('\n') if clean else []
        self.stdin_idx = 0

        for i, (op, arg) in enumerate(self.insns):
            if op == 'LABEL':
                if arg in self.labels:
                    raise VMError(f"Duplicate label: {arg}")
                self.labels[arg] = i

    def _num(self, v):
        if isinstance(v, int):
            return v
        if isinstance(v, float):
            return v
        s = str(v).strip()
        if not s:
            return 0
        try:
            return int(s) if '.' not in s else float(s)
        except ValueError:
            return 0

    def _str(self, v):
        if isinstance(v, float):
            return str(int(v)) if v == int(v) else f"{v:.6g}"
        return str(v)

    def _truthy(self, v):
        if isinstance(v, (int, float)):
            return v != 0
        return len(str(v)) > 0

    def push(self, v):
        self.stack.append(v)

    def pop(self):
        if not self.stack:
            raise VMError("Stack underflow")
        return self.stack.pop()

    def run(self):
        while self.pc < len(self.insns) and not self.halted:
            op, arg = self.insns[self.pc]
            self.pc += 1
            if op == 'LABEL':
                continue
            self.exec_count += 1

            if op == 'NOP':
                pass
            elif op == 'HALT':
                self.halted = True

            # Stack
            elif op == 'PUSH_INT':
                self.push(int(arg))
            elif op == 'PUSH_STR':
                self.push(arg if arg is not None else "")
            elif op == 'DUP':
                self.push(self.stack[-1])
            elif op == 'DROP':
                self.pop()
            elif op == 'SWAP':
                self.stack[-1], self.stack[-2] = self.stack[-2], self.stack[-1]

            # Arithmetic
            elif op == 'ADD':
                b, a = self._num(self.pop()), self._num(self.pop())
                r = a + b
                self.push(int(r) if isinstance(a, int) and isinstance(b, int) else r)
            elif op == 'SUB':
                b, a = self._num(self.pop()), self._num(self.pop())
                r = a - b
                self.push(int(r) if isinstance(a, int) and isinstance(b, int) else r)
            elif op == 'MUL':
                b, a = self._num(self.pop()), self._num(self.pop())
                r = a * b
                self.push(int(r) if isinstance(a, int) and isinstance(b, int) else r)
            elif op == 'DIV':
                b, a = self._num(self.pop()), self._num(self.pop())
                if b == 0:
                    raise VMError("Division by zero")
                if isinstance(a, int) and isinstance(b, int):
                    self.push(int(a / b))
                else:
                    self.push(float(a) / float(b))
            elif op == 'MOD':
                b, a = self._num(self.pop()), self._num(self.pop())
                if b == 0:
                    raise VMError("Modulo by zero")
                self.push(a % b)
            elif op == 'NEG':
                v = self._num(self.pop())
                self.push(-v)

            # Comparison
            elif op == 'CMP_EQ':
                b, a = self.pop(), self.pop()
                self.push(1 if a == b else 0)
            elif op == 'CMP_NE':
                b, a = self.pop(), self.pop()
                self.push(1 if a != b else 0)
            elif op == 'CMP_LT':
                b, a = self._num(self.pop()), self._num(self.pop())
                self.push(1 if a < b else 0)
            elif op == 'CMP_LE':
                b, a = self._num(self.pop()), self._num(self.pop())
                self.push(1 if a <= b else 0)
            elif op == 'CMP_GT':
                b, a = self._num(self.pop()), self._num(self.pop())
                self.push(1 if a > b else 0)
            elif op == 'CMP_GE':
                b, a = self._num(self.pop()), self._num(self.pop())
                self.push(1 if a >= b else 0)

            # Logic
            elif op == 'NOT':
                self.push(0 if self._truthy(self.pop()) else 1)

            # Variables
            elif op == 'LOAD':
                self.push(self.vars.get(arg, 0))
            elif op == 'STORE':
                self.vars[arg] = self.pop()

            # Control flow
            elif op == 'JUMP':
                if arg not in self.labels:
                    raise VMError(f"Unknown label: {arg}")
                self.pc = self.labels[arg]
            elif op == 'JUMP_TRUE':
                if self._truthy(self.pop()):
                    self.pc = self.labels[arg]
            elif op == 'JUMP_FALSE':
                if not self._truthy(self.pop()):
                    self.pc = self.labels[arg]

            # I/O
            elif op == 'PRINT':
                self.output.append(self._str(self.pop()))
            elif op == 'PRINTLN':
                self.output.append(self._str(self.pop()) + '\n')
            elif op == 'READ_LINE':
                if self.stdin_idx < len(self.stdin_lines):
                    self.push(self.stdin_lines[self.stdin_idx])
                    self.stdin_idx += 1
                    self.push(1)
                else:
                    self.push("")
                    self.push(0)

            # Conversion
            elif op == 'TO_INT':
                self.push(int(self._num(self.pop())))
            elif op == 'TO_STR':
                self.push(self._str(self.pop()))

            # String
            elif op == 'CONCAT':
                b, a = self._str(self.pop()), self._str(self.pop())
                self.push(a + b)

            else:
                raise VMError(f"Unknown opcode: {op}")

        return ''.join(self.output)


def count_static_instructions(text):
    """Count non-empty, non-comment, non-LABEL lines."""
    count = 0
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith('#') or line.startswith('LABEL'):
            continue
        count += 1
    return count


def main():
    if len(sys.argv) < 2:
        print("Usage: python3 vm.py <program.asm> [< stdin]", file=sys.stderr)
        sys.exit(1)

    with open(sys.argv[1]) as f:
        text = f.read()

    stdin_data = ""
    if not sys.stdin.isatty():
        stdin_data = sys.stdin.read()

    program = parse_program(text)
    vm = VM(program, stdin_data)
    output = vm.run()
    sys.stdout.write(output)
    print(f"Static instructions: {count_static_instructions(text)}", file=sys.stderr)
    print(f"Dynamic instructions: {vm.exec_count}", file=sys.stderr)


if __name__ == '__main__':
    main()
